File: ios/Shared/Sources/Shared/DeviceContext.swift
//
//  DeviceContext.swift
//  Shared
//

import CryptoKit
import Foundation

/// Stable device ID (Keychain) and idempotency key generation. Used on ALL requests.
public enum DeviceContext {
    private static let lock = NSLock()
    private static var _deviceId: String?

    public static var deviceId: String {
        lock.lock()
        defer { lock.unlock() }
        if let id = _deviceId ?? KeychainHelper.getDeviceId() {
            _deviceId = id
            return id
        }
        let newId = UUID().uuidString
        _ = KeychainHelper.setDeviceId(newId)
        _deviceId = newId
        return newId
    }

    /// New UUID per write operation; persist for retries so retries use same key.
    public static func newIdempotencyKey() -> String {
        UUID().uuidString
    }

    /// Stable idempotency key for `POST /api/v1/devices/register` (same device + push token → same key for safe retries; new token → new key).
    public static func idempotencyKeyDeviceRegister(pushToken: String) -> String {
        let material = "\(deviceId)\u{1e}\(pushToken)"
        let digest = SHA256.hash(data: Data(material.utf8))
        let hex = digest.map { String(format: "%02x", $0) }.joined()
        return "device-register-\(hex)"
    }

    /// Stable idempotency key for `POST /api/v1/devices/unregister` (per device).
    public static func idempotencyKeyDeviceUnregister() -> String {
        let material = "\(deviceId)\u{1e}device-unregister"
        let digest = SHA256.hash(data: Data(material.utf8))
        let hex = digest.map { String(format: "%02x", $0) }.joined()
        return "device-unregister-\(hex)"
    }
}

File: android/shared/src/main/java/ai/aistroyka/shared/DeviceContext.kt
package ai.aistroyka.shared

import android.content.Context
import java.security.MessageDigest
import java.util.UUID

/**
 * Stable device id and idempotency keys (parity with iOS [DeviceContext]).
 */
object DeviceContext {
    private const val PREFS = "aistroyka_device"
    private const val KEY_DEVICE_ID = "device_id"

    private lateinit var appContext: Context

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    val deviceId: String
        get() {
            val c = appContext
            val p = c.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val existing = p.getString(KEY_DEVICE_ID, null)
            if (existing != null) return existing
            val id = UUID.randomUUID().toString()
            p.edit().putString(KEY_DEVICE_ID, id).apply()
            return id
        }

    fun newIdempotencyKey(): String = UUID.randomUUID().toString()

    /**
     * Stable idempotency key for `POST /api/v1/devices/register` (same device + push token → same key for safe retries).
     * Matches iOS [DeviceContext.idempotencyKeyDeviceRegister]: SHA-256 of `deviceId + RS + token`, prefix `device-register-`.
     */
    fun idempotencyKeyDeviceRegister(pushToken: String): String {
        val material = "${deviceId}\u001E$pushToken"
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(material.toByteArray(Charsets.UTF_8))
        val hex = digest.joinToString("") { b -> "%02x".format(b) }
        return "device-register-$hex"
    }

    /** Same scheme as iOS [DeviceContext.idempotencyKeyDeviceUnregister]. */
    fun idempotencyKeyDeviceUnregister(): String {
        val material = "${deviceId}\u001Edevice-unregister"
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(material.toByteArray(Charsets.UTF_8))
        val hex = digest.joinToString("") { b -> "%02x".format(b) }
        return "device-unregister-$hex"
    }
}
